src/engines/ws.ts

```diff
@@ -0,0 +1,205 @@
+import { WebSocket } from "isows";
+import {
+	ConnectionUnavailable,
+	EngineDisconnected,
+	ResponseError,
+	UnexpectedConnectionError,
+	UnexpectedServerResponse,
+} from "../errors";
+import { type RpcRequest, type RpcResponse, isLiveResult } from "../types";
+import { getIncrementalID } from "../util/getIncrementalID";
+import { retrieveRemoteVersion } from "../util/versionCheck";
+import {
+	AbstractEngine,
+	ConnectionStatus,
+	type EngineContext,
+	type EngineEvents,
+} from "./abstract";
+
+export class WebsocketEngine extends AbstractEngine {
+	private pinger?: Pinger;
+	private socket?: WebSocket;
+
+	constructor(context: EngineContext) {
+		super(context);
+		this.emitter.subscribe("disconnected", () => this.pinger?.stop());
+	}
+
+	private setStatus<T extends ConnectionStatus>(
+		status: T,
+		...args: EngineEvents[T]
+	) {
+		this.status = status;
+		this.emitter.emit(status, args);
+	}
+
+	private async requireStatus<T extends ConnectionStatus>(
+		status: T,
+	): Promise<true> {
+		if (this.status !== status) {
+			await this.emitter.subscribeOnce(status);
+		}
+
+		return true;
+	}
+
+	version(url: URL, timeout: number): Promise<string> {
+		return retrieveRemoteVersion(url, timeout);
+	}
+
+	async connect(url: URL): Promise<void> {
+		this.connection.url = url;
+		this.setStatus(ConnectionStatus.Connecting);
+		const socket = new WebSocket(url.toString(), "cbor");
+		const ready = new Promise<void>((resolve, reject) => {
+			socket.addEventListener("open", () => {
+				this.setStatus(ConnectionStatus.Connected);
+				resolve();
+			});
+
+			socket.addEventListener("error", (e) => {
+				const error = new UnexpectedConnectionError(
+					"error" in e ? e.error : "An unexpected error occurred",
+				);
+				this.setStatus(ConnectionStatus.Error, error);
+				reject(error);
+			});
+
+			socket.addEventListener("close", () => {
+				this.setStatus(ConnectionStatus.Disconnected);
+			});
+
+			socket.addEventListener("message", async ({ data }) => {
+				const decoded = this.decodeCbor(
+					data instanceof Blob
+						? await data.arrayBuffer()
```

Comment:
What will catch this if it fail? If nothing the process with crash on unhandled rejection

Reply:
Nothing, there's no way to point this back to in which place the request was originally made

Reply:
You should emit an error event in that case